### PR TITLE
fix: false positive dataframe-like check

### LINKED ISSUE #1602: DataFrameLike False Positive When __getattr__ Presents
### Describe the bug

When a class implemented `__getattr__` or `__getattribute__`, marimo will treat this instance as a `DataFrameLike` object, trying to render a table from it and raise exceptions.

### Environment

{
  "marimo": "0.6.17",
  "OS": "Darwin",
  "OS Version": "23.5.0",
  "Processor": "arm",
  "Python Version": "3.11.5",
  "Binaries": {
    "Browser": "125.0.6422.142",
    "Node": "v22.1.0"
  },
  "Requirements": {
    "click": "8.1.7",
    "importlib-resources": "missing",
    "jedi": "0.19.1",
    "markdown": "3.6",
    "pymdown-extensions": "10.8.1",
    "pygments": "2.18.0",
    "tomlkit": "0.12.5",
    "uvicorn": "0.30.1",
    "starlette": "0.37.2",
    "websockets": "12.0",
    "typing-extensions": "4.7.1",
    "black": "24.4.2"
  }
}

### Code to reproduce

```python
class Boom:
    def __getattr__(self, _):
        return ...
```

```python
b = Boom()
```

This will result an error:
`TypeError: 'ellipsis' object is not callable`

### PR DESCRIPTION
**Pull request by mscolnick**
Fixes #1602

This is a bit more robust checking if its "dataframe-like" and has the dataframe protocol

### CODE REVIEW

File: marimo/_plugins/ui/_impl/tables/types.py [hunk 1]
```diff
@@ -11,3 +11,11 @@ class DataFrameLike(Protocol):
     def __dataframe__(
         self, nan_as_null: bool = False, allow_copy: bool = True
     ) -> DataFrame: ...
+
+
+def is_dataframe_like(value: object) -> bool:
+    return (
+        isinstance(value, DataFrameLike)
+        and hasattr(value, "__dataframe__")
+        and callable(value.__dataframe__)
```
Comment (Somainer) [inline marimo/_plugins/ui/_impl/tables/types.py:20]: What if `__getattr__` itself returns callable?

File: marimo/_plugins/ui/_impl/tables/utils.py [hunk 2]
```diff
@@ -46,7 +48,7 @@ def get_table_manager_or_none(data: Any) -> TableManager[Any] | None:
                 return manager(data)
 
     # If we have a DataFrameLike object, use the DataFrameProtocolTableManager
-    if isinstance(data, DataFrameLike):
+    if is_dataframe_like(data):
         return DataFrameProtocolTableManager(data)
```
Comment (Somainer) [inline marimo/_plugins/ui/_impl/tables/utils.py:52]: Maybe we can add a fall back to normal rendering approach if `__dataframe__` method itself returns a non-dataframe object or raises an exception.

File: marimo/_plugins/ui/_impl/tables/types.py (hunk 1 above)
Comment (mscolnick) [reply marimo/_plugins/ui/_impl/tables/types.py:20]: added more defensiveness: checking `value.__dataframe__ is value.__dataframe__` which could guard against some (but not all) of `__getattr__` returning a callable.

also added a check inside `DataFrameProtocolTableManager` after constructing the dataframe to check its shape further and falling back to `None`.

File: marimo/_plugins/ui/_impl/tables/utils.py (hunk 2 above)
Comment (mscolnick) [reply marimo/_plugins/ui/_impl/tables/utils.py]: done, thanks for the suggestion!

**Pull request merged**